jmsSource/src/test/java/io/vantiq/extsrc/jmsSource/TestJMS.java

```diff
@@ -0,0 +1,640 @@
+        
+        // Query the source
+        queryResponse = vantiq.query(testSourceName, queryParams);
+        
+        // Should query successfully, but message should be null since it was unable to send
+        assert !queryResponse.hasErrors();
+        responseBody = (JsonObject) queryResponse.getBody();
+        responseMessageElement = responseBody.get("message");
+        assertTrue(responseMessageElement.isJsonNull());
+        
+        // Delete the Source from VANTIQ
+        deleteSource();
+        core.stop();  
+    }
+    
+    @Test
+    public void testHeadersAndProperties() {
+        checkAllJMSProperties(false);
+        
+        // Check that Source and Type do not already exist in namespace, and skip test if they do
+        assumeFalse(checkSourceExists());
+        
+        // Setup a VANTIQ JMS Source, and start running the core
+        setupSource(createSourceDef(false));
+        
+        // Create message to send
+        Date date = new Date();
+        String message = "A message sent at time: " + dateFormat.format(date);
+        
+        // Publish message to the source (send to queue)
+        Map<String,Object> sendToQueueParams = new LinkedHashMap<String,Object>();
+        Map<String, Object> headers = new LinkedHashMap<String, Object>();
+        Map<String, Object> properties = new LinkedHashMap<String, Object>();
+        Map<String, String> replyTo = new LinkedHashMap<String, String>();
+        
+        replyTo.put("queue", testJMSQueue);
+        headers.put("JMSReplyTo", replyTo);
+        headers.put("JMSType", "TextMessage");
+        headers.put("JMSCorrelationID", "someID");
+        properties.put("randomProperty1", "some stuff");
+        properties.put("randomProperty2", 1);
+        
+        sendToQueueParams.put("properties", properties);
+        sendToQueueParams.put("headers", headers);
+        sendToQueueParams.put("message", message);
+        sendToQueueParams.put("queue", testJMSQueue);
+        vantiq.publish("sources", testSourceName, sendToQueueParams);
+        
+        // Query with no operation set
+        Map<String,Object> queryParams = new LinkedHashMap<String,Object>();
+        queryParams.put("queue", testJMSQueue);
+        queryParams.put("operation", "read");
+        VantiqResponse queryResponse = vantiq.query(testSourceName, queryParams);
+        
+        // Should query successfully, and message/properties/headers/queue should all be present
+        assert !queryResponse.hasErrors();
+        JsonObject responseBody = (JsonObject) queryResponse.getBody();
+        String responseMessage = (String) responseBody.get("message").getAsString();
+        assert responseMessage.equals(message);
+        assert responseBody.get("headers") instanceof JsonObject;
+        JsonObject headersObject = responseBody.get("headers").getAsJsonObject();
+        assert testJMSQueue.contains(headersObject.get("JMSReplyTo").getAsString());
```

Comment:
The reason I wrote the test like this is because the value returned from `JMSReplyTo` is actually a truncated version of the `testJMSQueue` value, which is the full name used to create the queue and send messages.

So I am checking that the value we get back is contained inside the full name we use to actually create JMS Destinations (queues in this case).

Reply:
Ah, OK.  My misunderstanding